vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,787 @@
+  //    Balsubramani, Akshay, Sanjoy Dasgupta, and Yoav Freund. "The fast convergence of
+  //    incremental PCA." Advances in neural information processing systems 26 (2013).
+
+  auto weights = emt_router_random(exs, rng);
+
+  std::map<int, float> sums;
+  emt_feats means;
+
+  for (auto& ex : exs)
+  {
+    for (auto& p : ex) { sums[p.first] += p.second; }
+  }
+  for (auto& p : sums) { means.emplace_back(p.first, p.second / exs.size()); }
+
+  std::vector<emt_feats> centered_exs;
+  for (auto& e : exs) { centered_exs.push_back(emt_scale_add(1, e, -1, means)); }
+
+  int n_epochs = 40;  // the bigger the better eigen approximation
+
+  for (int i = 0; i < n_epochs; i++)
+  {
+    // reseting n on each epoch gives
+    // projectors that are substantially
+    // closer to the true top eigen vector
+    // in experiments
+    float n = 1;
+    emt_shuffle(centered_exs.begin(), centered_exs.end(), &rng);
+
+    for (emt_feats fs : centered_exs)
+    {
+      // in matrix multiplication notation this looks like:
+      // weights = weights + (1/n) * inner(outer(fs,fs), weights)
+      //         = weights + (1/n) * fs * inner(fs,weights)
+      //         =          weights+(1/n)*inner(fs,weights)*fs
+      weights = emt_scale_add(1, weights, (1 / n) * emt_inner(fs, weights), fs);
+      emt_normalize(weights);
+      n += 1;
+    }
+  }
+
+  return weights;
+}
+
+emt_feats emt_router(std::vector<emt_feats> exs, emt_router_type router_type, VW::rand_state& rng)
+{
+  if (router_type == emt_router_type::random) { return emt_router_random(exs, rng); }
+  else { return emt_router_eigen(exs, rng); }
+}
+////////////////////////////end of helper/////////////////////////
+//////////////////////////////////////////////////////////////////
+
+////////////////////////eigen_memory_tree///////////////////
+////////////////////////////////////////////////////////////
+emt_node* node_route(emt_node& cn, emt_example& ec)
+{
+  return emt_inner(ec.base, cn.router_weights) < cn.router_decision ? cn.left.get() : cn.right.get();
+}
+
+emt_node* tree_route(emt_tree& b, emt_example& ec)
+{
+  emt_node* cn = b.root.get();
+  while (cn->left != nullptr) { cn = node_route(*cn, ec); }
+  return cn;
+}
+
+void tree_bound(emt_tree& b, emt_example* ec)
+{
+  emt_example* to_delete = b.bounder->bound(ec);
+
+  if (to_delete == nullptr) { return; }
+
+  emt_node& cn = *tree_route(b, *to_delete);
+
+  for (auto iter = cn.examples.begin(); iter != cn.examples.end(); iter++)
+  {
+    if (iter->get() == to_delete)
+    {
+      cn.examples.erase(iter);
+      return;
+    }
+  }
+}
+
+float scorer_initial(const VW::example& ex) { return 1 - std::exp(-std::sqrt(ex.total_sum_feat_sq)); }
+
+void scorer_features(const emt_feats& f1, features& out)
+{
+  out.clear();
+  for (auto p : f1)
+  {
+    if (p.second != 0) { out.push_back(p.second, p.first); }
+  }
+}
+
+void scorer_example(emt_tree& b, const emt_example& ex1, const emt_example& ex2)
```

Comment:
just to confirm what kind of asymmetry i mean:
scorer_example({0: a, 1: b}, {0: c})  = some_f({0: a, 1: c, 2: b})
scorer_example({0:c}, {0: a, 1: b}) = some_f({0: c, 1: a, 3: b})

Reply:
Good catch. The asymmetry is not desired.  Here's what we really wanted...

scorer_example({0: a, 1: b}, {0: c}) = some_f({00:ac, 10:bc})

It that possible using vw's interactions. I think I did the offset because it seemed to me like VW wasn't calculating the interactions at all for features in different namespaces with identical indexes (which I know is a strange use case to begin with).

That is, users could say --interactions

Reply:
(also worth noting, before moving all this to VW we were manually calculating these features so this wasn't an issue)